File: src/main/java/org/excavator/grpc/GrpcClientApplication.java
package org.excavator.grpc;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

public class GrpcClientApplication {
    public static void main(String[] args) {
        ManagedChannel channel = ManagedChannelBuilder.forAddress("localhost", 53000)
                .usePlaintext().build();

        ProductGrpc.ProductBlockingStub client = ProductGrpc.newBlockingStub(channel);

        ProductReviewRequest review = ProductReviewRequest.newBuilder()
                .setReview("cmonkey")
                .setProductId("0123456789")
                .setReviewerEmail("42.codemonkey at gmail.com")
                .setFiveStarRating(5)
                .build();

        ProductReviewResponse result = client.createOrUpdateReview(review);

        System.out.println("result of posing review ; " + result.getStatus().name());

        ProductReviewRequest badLangReview = ProductReviewRequest.newBuilder()
                .setReview("F*ck product")
                .setProductId("012345678")
                .setReviewerEmail("42.codemonkey at gmail.com")
                .setFiveStarRating(5).build();

        result = client.createOrUpdateReview(badLangReview);

        System.out.println("result of posting review: " + result.getStatus().name());

        ProductReviewRequest invalidRatingReview = ProductReviewRequest.newBuilder()
                .setReview("Bad product!")
                .setProductId("EAN132069854")
                .setReviewerEmail("42.codemonkey at gmail.com")
                .setFiveStarRating(-5)
                .build();

        result = client.createOrUpdateReview(invalidRatingReview);

        System.out.println("Result of posting review: " + result.getStatus().name());
    }
}

File: src/main/scala/org/excavator/grpc/GrpcClientApplication.scala
package org.excavator.grpc

import java.util.concurrent.atomic.AtomicInteger

import io.grpc.ManagedChannelBuilder
import io.grpc.stub.StreamObserver
import org.slf4j.LoggerFactory

class GrpcClientApplication(host: String, port: Int) {

  val logger = LoggerFactory.getLogger(classOf[GrpcClientApplication])
  val atomicInteger = new AtomicInteger(0)

  var blockingClient: ProductGrpc.ProductBlockingStub = _
  var client: ProductGrpc.ProductStub = _

  connect()

  def connect() = {
    val channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext.build
    Runtime.getRuntime.addShutdownHook(new Thread() {
      override def run(): Unit =
        channel.shutdownNow()
    })
    blockingClient = ProductGrpc.newBlockingStub(channel)
    client = ProductGrpc.newStub(channel)
  }

  def createReview(request: ProductReviewRequest) = {
    val response = blockingClient.createOrUpdateReview(request)

    logger.info(s"createReview response = ${response}")

    response
  }

  def responseStream(request: ProductReviewRequest) = {
    client.getResponse(request, new StreamObserver[ProductReviewResponse] {
      override def onNext(value: ProductReviewResponse): Unit = {
        logger.info(s"next response = ${value.getStatus.name()}")
        atomicInteger.incrementAndGet()
      }

      override def onCompleted(): Unit = {
        logger.info(s"stream onCompleted and count = ${atomicInteger.getAndSet(0)}")
      }

      override def onError(t: Throwable): Unit = {
        logger.error(s"stream error = ${t}")
      }
    })
  }

}
